### PR TITLE
Phase 8: vigilo-config-migration

### PR DESCRIPTION
**Pull request by sh1ny**
## Summary

**Phase 8: Vigilo Config Migration**  
**Goal:** Maintainers can translate supported Vigilo/VDE YAML into strict Correlia config and get clear failures whenever Vigilo semantics cannot be preserved.  
**Status:** Verified ✓

Phase 8 delivers a standalone Vigilo/VDE-to-Correlia migration path plus runtime topology capture-group support. The implementation converts supported rules, topology, and email output configuration into strict Correlia YAML, rejects unsupported Vigilo semantics before reporting success, and validates staged output before atomic promotion.

### Plan 08-01: Runtime topology capture-group support
Hostname-derived topology tags now flow from strict YAML schema through compiled config into runtime enrichment, with literal tags applied first and capture-group substitutions bounded and diagnosed consistently.

**Key files:**
- `app/config/topology.py — schema, compiled rule, and loader validation`
- `app/processing/enrichment.py — hostname regex match capture and derived tag application`
- `tests/test_topology_enrichment.py — acceptance/rejection and enrichment regression tests`

### Plan 08-02: Vigilo/VDE migration CLI
Standalone Vigilo/VDE-to-Correlia migration CLI that fails closed on unsupported semantics, rejects plaintext SMTP credentials, and atomically promotes only loader-validated config files.

**Key files:**
- `scripts/migrate_vigilo_config.py`
- `tests/test_vigilo_config_migration.py`
- `tests/fixtures/vigilo/rules_valid.yaml`
- `tests/fixtures/vigilo/topology_valid.yaml`
- `tests/fixtures/vigilo/plugins_valid.yaml`
- `tests/fixtures/vigilo/plugins_with_unknown_email_option.yaml`
- `tests/fixtures/vigilo/plugins_with_unknown_section.yaml`
- `tests/fixtures/vigilo/plugins_with_no_outputs.yaml`
- `CONFIGURATION.md`

### Plan 08-03: Gap closure for placeholder rejection and rollback
Fail-closed plugin placeholder rejection and atomic promotion rollback for the Vigilo config migration CLI

**Key files:**
- `tests/fixtures/vigilo/plugins_with_placeholder_option.yaml`
- `scripts/migrate_vigilo_config.py`
- `tests/test_vigilo_config_migration.py`

## Requirements Addressed

- `CFG-01`
- `CFG-02`
- `CFG-03`
- `CFG-04`
- `CFG-05`
- `CFG-06`
- `CFG-07`

## Verification

- [x] Automated verification: `08-VERIFICATION.md` reports `status: passed` and ends with `## VERIFICATION PASSED`.
- [x] Full gate recorded in verification: `make lint`, `make typecheck`, and `make test` (533 tests).
- [x] Human verification required: none.

## Key Decisions

- [Phase 08]: tag_capture_groups is a strict dict[str, int] with Field(default_factory=dict), so existing literal-only topology YAML remains valid. — Decision recorded during execution of 08-01-PLAN.md.
- [Phase 08]: Derived capture-group tags are applied after literal tags; a derived value that overrides a literal value updates tags_added to the final captured value and records literal->derived in conflicts. — Decision recorded during execution of 08-01-PLAN.md.
- [Phase 08]: Subnet topology rules intentionally reject tag_capture_groups via Pydantic extra=forbid, preserving D-11 literal-only semantics. — Decision recorded during execution of 08-01-PLAN.md.

## Additional Repository Maintenance

This branch contains prior v1.1 planning and execution history in addition to the Phase 8 artifacts because the local phase branch is ahead of `origin/main` by the full milestone workstream range. Reviewers should expect ROADMAP/STATE and prior phase planning artifacts in the diff; Phase 8 verification remains the shipping gate for this PR.

## User Stories & Acceptance Criteria

- Maintainers can run the migration CLI against supported Vigilo/VDE YAML and receive strict Correlia YAML outputs.
- Unsupported Vigilo semantics fail with clear diagnostics instead of silent lossy conversion.
- Acceptance criteria are covered by CFG-01..CFG-07 and the Phase 8 verification report.

## Risks & Dependencies

- Depends on Phase 5 security and HTTP controls, as recorded in ROADMAP.
- Main rollout risk is operator review of generated config before production use; the migration CLI validates output but does not infer unsupported Vigilo semantics.

## Success Metrics & Release Criteria

- Release when automated verification and required manual checks pass.
- Phase 8 verification passed with lint, typecheck, and full test gate recorded.

## Stakeholder Review & Approval

- Product owner approval pending for vigilo-config-migration.

## TDD Audit

| Test commit | Impl commit | gate_status |
|---|---|---|
| `2bb4a7c` docs(07): capture phase context | — | missing |
| `8c063e9` docs(state): record phase 7 context session | — | missing |
| `bd09733` docs(07): make decision summary fields nullable for no-op events | — | missing |
| `2024b85` docs(07): research phase domain — incident event audit trail | — | missing |
| `f41fed7` docs(07): plan 07-01 audit foundation | — | missing |
| `91f7594` docs(07): plan 07-02 ingress refactor | — | missing |
| `579ae19` docs(07): plan 07-03 audit read API | — | missing |
| `052535b` docs(07): plan outline | — | missing |
| `1b28c59` docs(07): pattern map | — | missing |
| `8d9f95d` docs(07): revise 07-01 for D-08 read-time redaction | — | missing |
| `d96db54` docs(07): revise 07-03 for D-08 read-time redaction | — | missing |
| `451c076` docs(07): mark phase 7 planned | — | missing |
| `5b97eb4` docs(07): finalize planning state and roadmap | — | missing |
| `233bdb5` docs(07): begin phase execution | — | missing |
| `e52cad8` docs(07): fix stopped_at for execution | — | missing |
| `81c5329` docs(07): checkpoint before executor dispatch | — | missing |
| `b8cd910` docs(07): checkpoint wave 1 executing 07-01 | — | missing |
| `a2f2c21` test(07-01): add failing audit schema, settings, and callsite tests | `2a53350` feat(07-01): add incident_events schema, ORM model, and audit settings | missing |
| `1320edb` docs(07): checkpoint wave 1 task 1 complete | — | missing |
| `ec98f48` fix(07-01): tighten raw_payload CHECK to object-only per D-04/D-07 | — | missing |
| `2334e67` feat(07-01): add audit domain contracts, redaction, HMAC, and cursor helpers | — | missing |
| `d2cf4e5` docs(07): checkpoint wave 1 task 2 complete | — | missing |
| `d03de68` feat(07-01): add audit persistence repository and tests | — | missing |
| `e6e3dfb` docs(07): Wave 1 closeout — 07-01-SUMMARY.md, STATE, ROADMAP | — | missing |
| `2e294f3` docs(07): remove duplicate Wave 2 block from ROADMAP | — | missing |
| `32f4bed` refactor(07-02): move commit ownership from managers to ingress | — | missing |
| `3e0bc92` refactor(07-02): write one audit row per accepted ingress transaction | — | missing |
| `99431a2` test(07-02): prove audit remains observational on operator and sweep paths | `d4dc915` fix(07-02): remove stale fake-session references from summary and fix mypy type errors | missing |
| `c0384b5` docs(07-02): update ROADMAP.md and STATE.md for plan completion | — | missing |
| `fab328f` docs(07-02): update summary bookkeeping — add audit.py and test_domain_audit.py to files modified, bump deviation count to 6 | — | missing |
| `621d74f` feat(07-03): create read-only incident-events operator router | — | missing |
| `0a4d6a2` feat(07-03): register audit router and classify route as operator | — | missing |
| `3773611` test(07-03): add audit API, rate-limit, and ingress correlation tests | `ad79534` fix(07-03): omit sensitive tag keys from audit read responses (D-08/D-15) | missing |
| `4cb049e` docs(07-03): complete Phase 7 — summary, STATE.md, ROADMAP.md | — | missing |
| `5e60eb0` docs(07): add Phase 7 code review (warnings only, no blockers) | — | missing |
| `2f5c552` docs(07): add Phase 7 verification report and refine review findings | — | missing |
| `33928ef` docs(07): mark Phase 7 complete in STATE.md and REQUIREMENTS.md | — | missing |
| `66e8085` docs(phase-7): evolve PROJECT.md after phase completion | — | missing |
| `7a335c6` docs(08): capture phase context | — | missing |
| `7302936` docs(state): record phase 8 context session | — | missing |
| `4636659` docs(08): research Vigilo config migration domain | — | missing |
| `2317368` docs(08): create vigilo config migration plans | — | missing |
| `d3f2251` docs(08): fix topology plan task order | — | missing |
| `14ad1cb` docs(08): clarify migration plan test verification | — | missing |
| `8447cd5` docs(08): address phase plan check findings | — | missing |
| `6ac0c18` docs(08): plan Phase 8 Vigilo config migration (accepted with known issues) | — | missing |
| `1b3aed5` docs(08): add executor handoff note for accepted planning risks | — | missing |
| `faa7ccc` feat(08-01): extend topology schema and loader for tag_capture_groups | — | missing |
| `c29b394` feat(08-01): apply derived hostname tags from capture groups | — | missing |
| `7a5372a` test(08-01): run topology capture-group regression | `c1d94b0` feat(08-02): implement Vigilo/VDE-to-Correlia migration CLI | missing |
| `46fd0a4` docs(08): add continue-here handoff for in-flight Plan 2 execution | — | missing |
| `373322f` docs(08-01): complete runtime topology capture-group plan | — | missing |
| `8f12b20` test(08-02): add migration fixtures and CLI tests | `46df9b9` fix(08-02): reject rule actions that reference missing output plugins | missing |
| `a03b71d` docs(08-02): update CONFIGURATION.md for migration CLI | — | missing |
| `2af8484` docs(08-02): complete Phase 8 Plan 2 summary and update planning state | — | missing |
| `5deca5c` fix(08-02): add noqa for sys.path-inserted app imports | — | missing |
| `bafd44e` docs(08-02): record post-execution lint fix in summary | — | missing |
| `860f495` docs(08-03): gap-closure plan for D-05/D-15 | — | missing |
| `30b2268` test(08-03): add failing D-05 placeholder rejection regression test and fixture | `acd6ed0` feat(08-03): reject D-05 plugin option placeholders in preflight and transform | missing |
| `a66ed40` test(08-03): add failing D-15 mid-promotion rollback regression test | `86ceb27` feat(08-03): roll back newly-promoted files on D-15 mid-promotion failure | missing |
| `cfe8136` fix(08-03): scan raw config values in transform guard and import Iterator | — | missing |
| `0ef4bfb` docs(08-03): add gap-closure summary for D-05/D-15 | — | missing |
| `5d253de` fix(08-03): close D-05 substring placeholders and D-14 out_dir creation gaps | — | missing |
| `1dbe00b` fix(08-03): verifier-hardened D-14 rollback and mkdtemp edge case | — | missing |
| `1ffa5e8` fix(08-03): reject placeholder syntax in Vigilo output names; add regression test | — | missing |
| `78f5928` docs(08): record passing Phase 8 verification | — | missing |
| `91a2a0b` docs(08): mark Phase 8 verified complete | — | missing |
| `c5a7bb3` docs(08): refresh verification report for shipping | — | missing |
| `e79fdf3` docs(08): ship phase 8 via PR #3 | — | missing |

Aggregate: 0 skill, 0 fallback, 0 exempt — 69 missing.

gate_status: skill=0, fallback=0, exempt=0, missing=69

## Summary by CodeRabbit

* **New Features**
  * Added operator-facing **Incident Events** API (`GET /v1/incident-events`) with filtering, cursor pagination, and bounded/redacted results.
  * Introduced an **append-only incident audit trail** with configurable raw-payload size limits and required HMAC validation.
  * Added **topology capture groups** to support derived hostname tags.
  * Added a **Vigilo/VDE migration CLI** to generate validated Correlia YAML outputs (`rules.yaml`, `topology.yaml`, `plugins.yaml`).
  * Added new configuration for audit raw payload max bytes and HMAC key.
* **Bug Fixes**
  * Strengthened input validation (including cursor and `source_id` length) and improved audit behavior to avoid writing during non-ingress operations.
* **Documentation**
  * Updated configuration and migration examples for the latest workflow.

### CODE REVIEW

File: scripts/migrate_vigilo_config.py [hunk 1]
```diff
@@ -0,0 +1,1312 @@
...
+        for issue in issues:
+            print(f"  [{issue.code}] {issue.message}", file=sys.stderr)
+        return 1
+
+    raw_rules = _load_yaml(rules_path)
+    raw_topology = _load_yaml(topology_path)
+    raw_plugins = _load_yaml(plugins_path)
+
+    rules_list = raw_rules.get("rules", []) if isinstance(raw_rules, dict) else []
+    topology_rules = raw_topology.get("topology_rules", {}) if isinstance(raw_topology, dict) else {}
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py]: > `rules or topology YAML is not a mapping`

Addressed: rules and topology inputs now require mapping documents and their required top-level wrappers before migration can proceed.

File: scripts/migrate_vigilo_config.py [hunk 2]
```diff
@@ -0,0 +1,1312 @@
...
+        print("Migration failed:", file=sys.stderr)
+        for issue in issues:
+            print(f"  [{issue.code}] {issue.location}: {issue.message}", file=sys.stderr)
+        return 1
+
+    transformed_rules = _transform_rules(rules_list)
+    transformed_topology = _transform_topology(topology_rules)
+    transformed_plugins = _transform_plugins(raw_plugins)
+
+    staging = tempfile.mkdtemp(prefix="migrate_staging_", dir=out_dir.parent)
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py]: > `mkdtemp(..., dir=out_dir.parent)` runs before `_promote()` can call `out_dir.mkdir(parents=True)`

Addressed: staging now uses the nearest existing ancestor, so nested output parents are created only during promotion after staging validation succeeds.

File: .planning/phases/07-incident-event-audit-trail/07-CONTEXT.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+Phase 7 adds an append-only `incident_events` audit table that records every accepted normalized source event, a redacted snapshot of the raw source payload, and a bounded decision summary. Audit rows are used for operator traceability only; they must not influence incident aggregation, notification thresholds, recovery, or lifecycle transitions. Operators can browse audit rows through a canonical read-only endpoint and correlate them to incidents when one exists.
+
+Phase 7 does not implement notification plugin delivery confirmation, long-term raw event warehousing, webhook endpoint compatibility, Vigilo config migration, output-plugin hardening, or deployment packaging. It also does not allow summary mutation of incidents.
+
+</domain>
+
+<decisions>
+## Implementation Decisions
+
+### Audit Write Timing
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply .planning/phases/07-incident-event-audit-trail/07-CONTEXT.md]: > `Keep this section at ##`

Addressed: promoted `Audit Write Timing` to `##`, matching the requested heading level.

**Comment by sh1ny** [comment]
@codex review

File: app/processing/ingress.py [hunk 4]
```diff
@@ -186,49 +313,244 @@ async def process_payload(
...
-                task_runner=self._task_runner,
-                plugin_registry=self._plugin_registry,
-                config_hash=self._config_hash,
+    async def _submit_notifications(
+        self,
+        incident_id: Any,
+        decision: RuleDecision | None,
+    ) -> tuple[NotificationResult, ...]:
+        if self._task_runner is None or decision is None:
+            return ()
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply app/processing/ingress.py]: > When a processor is built with rules/actions but without a task runner ... return a dispatch_failed result per action or fail fast instead.

Addressed: configured actions now return deterministic `dispatch_failed` results when no task runner is available; each records one failed submission attempt. A regression test also preserves the empty result for no rule decision.

**Comment by sh1ny** [comment]
@codex review

File: scripts/migrate_vigilo_config.py [hunk 5]
```diff
@@ -0,0 +1,1405 @@
...
+            issues.append(
+                MigrationIssue(
+                    domain="rules",
+                    location=f"match.tags[{key}]",
+                    code="unsupported_tag_wildcard",
+                    message=f"tag value for '{key}' contains unsupported fnmatch metacharacters",
+                    requirement="CFG-06",
+                )
+            )
+            continue
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py:283]: > regex-style tag matching such as `datacenter: ".+"`

Not addressing: Vigilo matches tag values with `fnmatch.fnmatch`, where `.+` has no wildcard meaning and matches only the literal `.+`; preserving it as an exact Correlia tag match retains source behavior. The migration already rejects actual fnmatch metacharacters (`*`, `?`, `[`, `]`).

File: scripts/migrate_vigilo_config.py [hunk 6]
```diff
@@ -0,0 +1,1405 @@
...
+                    code="unsupported_plugin_class",
+                    message=f"source class '{class_name}' cannot be mapped to {_EMAIL_TARGET_CLASS}",
+                    requirement="CFG-06",
+                )
+            )
+            continue
+
+        config = output.get("config", {})
+        if not isinstance(config, dict):
+            config = {}
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py]: > This should be a preflight error instead of falling back to defaults when the source config is present but not a mapping.

Addressed: Present non-mapping email output configs now produce a structured `unsupported_plugin_option` error, and direct transforms fail instead of substituting delivery defaults.

File: scripts/migrate_vigilo_config.py [hunk 7]
```diff
@@ -0,0 +1,1405 @@
...
+                        location=f"{loc}.match.severity",
+                        code="invalid_rule_severity",
+                        message="match.severity must be a non-empty list",
+                        requirement="CFG-06",
+                    )
+                )
+
+            tags = match_block.get("tags", {})
+            if not isinstance(tags, dict):
+                tags = {}
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py]: > That broadens the migrated rule instead of failing closed on semantics the converter cannot preserve.

Addressed: Non-mapping `match.tags` now produces `invalid_rule_tags` during preflight, and direct transforms reject it instead of dropping tag constraints.

File: app/domain/audit.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+    id: UUID
+    accepted_at: datetime
+    event_timestamp: datetime
+    source_id: BoundedString
+    fingerprint: BoundedString
+    event_type: EventType
+    severity: Severity
+    host: BoundedString
+    service: BoundedString | None
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply app/domain/audit.py:139]: > reject or truncate these fields at normalization/write time, or widen the read contract.

Addressed: Icinga request validation now caps `host` and `service` at 256 characters, returning 422 before ingress or audit persistence.

**Comment by sh1ny** [comment]
@codex review

File: scripts/migrate_vigilo_config.py [hunk 9]
```diff
@@ -0,0 +1,1424 @@
...
+    )
+    output_rules: list[dict[str, Any]] = []
+    for new_priority, rule in enumerate(sorted_rules, start=1):
+        match_block = rule.get("match", {})
+        if not isinstance(match_block, dict):
+            match_block = {}
+
+        host_pattern = match_block.get("host", "*")
+        if not isinstance(host_pattern, str):
+            host_pattern = "*"
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py]: > this should be a preflight error rather than defaulting unless the field is actually absent.

Addressed: Present non-string `match.host` values and present non-null non-string `match.service` values now fail preflight with `invalid_rule_matcher`; direct transforms reject them too. Omitted fields keep their defaults.

File: scripts/migrate_vigilo_config.py [hunk 10]
```diff
@@ -0,0 +1,1424 @@
...
+                code="unsupported_hostname_topology",
+                message="topology_rules must be a mapping",
+                requirement="CFG-06",
+            )
+        )
+        return issues
+
+    hostname_patterns = raw_topology.get("hostname_patterns", [])
+    if not isinstance(hostname_patterns, list):
+        hostname_patterns = []
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py:705]: > these malformed topology sections should be reported as CFG-06 input errors.

Addressed: Present non-list `hostname_patterns` and `ip_subnets` now return `invalid_topology_section` before output. Direct transforms reject those shapes; omitted sections remain empty by default.

File: scripts/migrate_vigilo_config.py [hunk 11]
```diff
@@ -0,0 +1,1424 @@
...
+                    message="priority must be an integer",
+                    requirement="CFG-06",
+                )
+            )
+        else:
+            priorities.append(priority)
+
+        raw_group_by = window.get("group_by", [])
+        if isinstance(raw_group_by, list):
+            _, group_issues = _rewrite_group_by(raw_group_by)
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py:533]: > Validate each list entry and emit `unsupported_group_by` for non-strings before calling the rewriter.

Addressed: Group-by rewriting now emits indexed `unsupported_group_by` issues for non-string entries, and direct rule transformation rejects those entries instead of reaching the prior `AttributeError` path.

File: app/processing/ingress.py [hunk 12]
```diff
@@ -186,49 +313,254 @@ async def process_payload(
...
+        incident_effect="none",
+        rule_name=None,
+        group_key=None,
+        incident_ids=(),
+        affected_incident_count=0,
+        incident_ids_truncated=False,
+        decision_reason=noop_reason or "no matching rule",
+        no_dispatch_reason=(noop_reason or "no_matching_rule")
+        if not matched_rules
+        else None,
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply app/processing/ingress.py]: > Use a bounded reason code here, or truncate/map the human-readable text separately.

Addressed: The free-form missing-group-by no-op family now maps to `missing_required_group_by_field` in the audit summary. The ingress decision retains its human-readable reason, while both persisted audit reason fields remain bounded.

File: app/config/settings.py [hunk 13]
```diff
@@ -49,6 +49,10 @@ class Settings(BaseSettings):
     rate_limit_window_seconds_health: int = Field(default=60, ge=1)
     rate_limit_sweep_interval_seconds: int = Field(default=60, ge=1, le=3600)
 
+    # Audit-trail settings (Phase 7)
+    audit_raw_payload_max_bytes: int = Field(default=65_536, ge=1_024, le=1_048_576)
+    audit_raw_payload_hmac_key: SecretStr
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply app/config/settings.py:54]: > please add the new secret to the documented environment variables.

Addressed: `CONFIGURATION.md` now documents required startup configuration for `CORRELIA_AUDIT_RAW_PAYLOAD_HMAC_KEY`, its audit-integrity purpose, and secret-manager-only provisioning.

File: app/processing/ingress.py [hunk 14]
```diff
@@ -129,25 +160,121 @@ async def process_payload(
...
+            await insert_incident_event(
+                session,
+                event_timestamp=event.timestamp,
+                source_id=event.source_id,
+                fingerprint=event.fingerprint,
+                event_type=event.event_type.value,
+                severity=event.severity.value,
+                host=event.host,
+                service=event.service,
+                incident_ids=[str(uid) for uid in summary.incident_ids],
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply app/processing/ingress.py]: > persist the full `lifecycle_result.incident_ids` for correlation and cap only the response/decision summary.

Addressed: recovery rows now persist all lifecycle incident IDs, while the decision summary and API projection remain capped at 20. Added regression coverage for 21-ID recovery persistence and 21st-ID filtering.

File: scripts/migrate_vigilo_config.py [hunk 15]
```diff
@@ -0,0 +1,1480 @@
...
+        target_tag = entry.get("target_tag")
+        tags = entry.get("tags", {})
+        if not isinstance(tags, dict):
+            tags = {}
+
+        rewritten_tags = {}
+        for key, value in tags.items():
+            new_key = _to_correlia_tag_key(str(key))
+            if new_key is not None:
+                rewritten_tags[new_key] = str(value)
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py:964]: > Reject non-string topology tag values during preflight rather than preserving typo-derived strings.

Addressed: preflight now rejects non-string literal hostname tag values and subnet values as structured CFG-06 input errors before transforms can stringify them.

File: scripts/migrate_vigilo_config.py [hunk 16]
```diff
@@ -0,0 +1,1512 @@
...
+        if isinstance(entry, dict)
+    ]
+    subnet_ids = _unique_slugs(subnet_names)
+
+    subnet_rules: list[dict[str, Any]] = []
+    for idx, entry in enumerate(ip_subnets):
+        if not isinstance(entry, dict):
+            continue
+        rule_id = subnet_ids[idx]
+        target_tag = str(entry.get("target_tag", ""))
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py:1000]: > Reject null topology target_tag values.

Addressed: preflight now rejects non-string `target_tag` values for capture-group hostname and subnet rules before transformation, while preserving missing-field and valid tag-key diagnostics.

**Comment by sh1ny** [comment]
> Docstring coverage is 12.11% which is insufficient. Write docstrings for the functions missing them to satisfy the coverage threshold.

Declined: Correlia does not define an 80% docstring-coverage gate in `pyproject.toml`, `Makefile`, or `.coderabbit.yaml`; applying that external threshold would create broad boilerplate churn across unrelated Phase 7/8 code. The repository's required lint, typecheck, and test gates pass, so this PR keeps documentation changes scoped to its behavior.

File: scripts/migrate_vigilo_config.py [hunk 17]
```diff
@@ -0,0 +1,1532 @@
...
+                MigrationIssue(
+                    domain="rules",
+                    location=f"match.tags[{key}]",
+                    code="unsupported_tag_wildcard",
+                    message=f"tag value for '{key}' contains unsupported fnmatch metacharacters",
+                    requirement="CFG-06",
+                )
+            )
+            continue
+        new_key = _to_correlia_tag_key(key)
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py:284]: > Reject non-string rule tag keys before rewriting.

Addressed: `_rewrite_match_tags` now rejects non-string keys before any string operation and emits the existing `invalid_topology_tag_key` CFG-06 issue with the key-specific location and message. Added CLI regressions for numeric and boolean YAML keys; both confirm a clean nonzero exit, no traceback, and no output promotion.

File: scripts/migrate_vigilo_config.py [hunk 18]
```diff
@@ -0,0 +1,1532 @@
...
+            issues.append(
+                MigrationIssue(
+                    domain="rules",
+                    location=f"{loc}.priority",
+                    code="invalid_rule_priority",
+                    message="priority is required and must be an integer",
+                    requirement="CFG-06",
+                )
+            )
+        elif not isinstance(priority, int):
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py]: > Reject boolean rule priorities explicitly.

Addressed: `_preflight_rules` now treats `bool` as invalid before accepting integer priorities. Added CLI regressions for both `True` and `False`; each now returns the structured `invalid_rule_priority` CFG-06 report without output promotion or a traceback.

File: scripts/migrate_vigilo_config.py [hunk 19]
```diff
@@ -0,0 +1,1543 @@
...
+                MigrationIssue(
+                    domain="topology",
+                    location=loc,
+                    code="unsupported_hostname_topology",
+                    message="hostname pattern entry must be a mapping",
+                    requirement="CFG-06",
+                )
+            )
+            continue
+        regex = entry.get("regex", "")
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py:719]: > Reject missing hostname regexes before emitting rules

Addressed: hostname preflight now rejects both missing and explicitly empty regexes with a CFG-06 issue at the entry’s `.regex` location before rule emission.

File: scripts/migrate_vigilo_config.py [hunk 20]
```diff
@@ -0,0 +1,1543 @@
...
+                issues.append(
+                    MigrationIssue(
+                        domain="topology",
+                        location=tag_location,
+                        code="invalid_topology_value",
+                        message="hostname pattern literal tag values must be strings",
+                        requirement="CFG-06",
+                    )
+                )
+            if _to_correlia_tag_key(str(tag_name)) is None:
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py]: > Reject non-string topology tag names before rewriting

Addressed: hostname literal tag names must now be strings; preflight reports a CFG-06 tag-key issue before key conversion and never validates a stringified source key.

File: scripts/migrate_vigilo_config.py [hunk 21]
```diff
@@ -0,0 +1,1575 @@
...
+                    MigrationIssue(
+                        domain="topology",
+                        location=tag_location,
+                        code="invalid_topology_tag_key",
+                        message="hostname pattern literal tag names must be strings",
+                        requirement="CFG-06",
+                    )
+                )
+                continue
+            if not isinstance(tag_value, str):
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply scripts/migrate_vigilo_config.py:820]: > “Apply the same `TagValue` min/max bounds here, and to subnet `value` entries below, before promotion.”

Addressed in `78e1a0f`: topology preflight now validates literal hostname-tag and subnet values through the canonical `TagValue` contract, rejecting empty and over-256-character values under CFG-06 before output promotion. Added CLI regressions for both boundaries.

File: app/config/topology.py [hunk 22]
```diff
@@ -25,6 +26,21 @@ def _tags_must_start_with_topology(cls, value: dict[str, str]) -> dict[str, str]
                 raise ValueError(f"tag key must start with 'topology.': {key}")
         return value
 
+    @field_validator("tag_capture_groups")
+    @classmethod
+    def _capture_groups_must_be_valid(
+        cls, value: dict[str, int]
+    ) -> dict[str, int]:
+        for key, group in value.items():
+            if not key.startswith("topology."):
```
[the remark this answers is not in the document]
Comment (sh1ny) [reply app/config/topology.py:37]: > “Validate capture-group keys against the same tag-key pattern used for event tags, not just the `topology.` prefix.”

Addressed in `78e1a0f`: `tag_capture_groups` now uses the canonical `TagKey` contract while retaining its `topology.` namespace and capture-group index validation. A regression rejects `topology.Datacenter` during configuration loading.
